engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+	if err != nil {
+		return errwrap.Wrapf(err, "failed serverupdate request")
+	}
+	obj.server = server
+	return nil
+}
+
+// Cmp compares two HetznerVMRes structs. Returns nil if the comparison holds
+// true, and otherwise an error is thrown to identify the difference.
+func (obj *HetznerVMRes) Cmp(r engine.Res) error {
+	// check if empty
+	if obj == nil && r == nil {
+		return nil
+	}
+	if (obj == nil) != (r == nil) {
+		return fmt.Errorf("one resource is empty")
+	}
+	// compare types
+	res, ok := r.(*HetznerVMRes)
+	if !ok {
+		return fmt.Errorf("not a %s", obj.Kind())
+	}
+	// compare resource fields
+	if obj.State != res.State {
+		return fmt.Errorf("state differs")
+	}
+	if obj.APIToken != res.APIToken {
+		return fmt.Errorf("apitoken differs")
+	}
+	if obj.ServerType != res.ServerType {
+		return fmt.Errorf("servertype differs")
+	}
+	if obj.Datacenter != res.Datacenter {
+		return fmt.Errorf("datacenter differs")
+	}
+	if obj.Image != res.Image {
+		return fmt.Errorf("image differs")
+	}
+	if obj.ServerRescueMode != res.ServerRescueMode {
+		return fmt.Errorf("serverrescuemode differs")
+	}
+	for i, key := range obj.ServerRescueSSHKeys {
+		if key != res.ServerRescueSSHKeys[i] {
+			return fmt.Errorf("serverrescuekeys differs")
+		}
+	}
+	return nil
+}
+
+// cmpServerSpecs compares the server specifications between the virtual
+// HetznerVMRes resource and its corresponding Hetzner instance. Returns true if
+// ServerType, Datacenter and Image match. Returns an error is the server is
+// unavailable (`absent`).
+func (obj HetznerVMRes) cmpServerSpecs() (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("cmpServerSpecs()")
+	}
+	if obj.server == nil {
+		return false, fmt.Errorf("server is unavailable")
+	}
+	if obj.ServerType != obj.server.ServerType.Name {
+		return false, nil
+	}
+	if obj.Datacenter != obj.server.Datacenter.Name {
+		return false, nil
+	}
+	if obj.Image != obj.server.Image.Name {
+		return false, nil
+	}
+	return true, nil
+}
+
+// createServer checks if the servername is not already in use, builds the
+// serverconfig in hcloud-go format from resource params, requests a server
+// creation with that configuration, and waits until the creation is confirmed.
+// Errors occur when the server exists already, the client fails, or the wait
+// step exits early (context cancellation, client failure or timeout).
+// TODO: add createServer permission field
+func (obj *HetznerVMRes) createServer(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("createServer()")
+	}
+	if obj.server != nil {
+		return fmt.Errorf("server already exists")
+	}
+	if err := obj.getServerConfig(ctx); err != nil {
+		return errwrap.Wrapf(err, "getServerConfig failed")
+	}
+	if _, _, err := obj.client.Server.Create(ctx, obj.serverconfig); err != nil {
+		return errwrap.Wrapf(err, "client.server.create failed")
+	}
+	if err := obj.waitFor(ctx, serverExists); err != nil {
```

Comment:
this should be obj.serverExists instead of course... And similarly for the other function args to waitFor.